File: flutter_app/lib/locale_store.dart
import 'dart:ui' as ui;
import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'l10n/app_locale.dart';
import 'l10n/strings.dart';

/// Til holati — profil ekranidan tanlanadi, qurilmada saqlanadi. Birinchi
/// marta ochilganda (hali qo'lda tanlanmagan bo'lsa) qurilma tiliga qarab
/// avtomatik aniqlanadi — qo'llab-quvvatlanmasa "O'zbekcha"ga qaytadi.
class LocaleStore extends ChangeNotifier {
  static const _prefKey = 'fp.locale';

  String _code = defaultLocaleCode;
  String get code => _code;

  Future<void> init() async {
    final prefs = await SharedPreferences.getInstance();
    final saved = prefs.getString(_prefKey);
    _code = saved ?? _detectDeviceLocale();
    notifyListeners();
  }

  String _detectDeviceLocale() {
    for (final deviceLocale in ui.PlatformDispatcher.instance.locales) {
      for (final supported in supportedLocales) {
        if (supported.code == deviceLocale.languageCode) return supported.code;
      }
    }
    return defaultLocaleCode;
  }

  Future<void> setLocale(String code) async {
    _code = code;
    notifyListeners();
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_prefKey, code);
  }

  String t(String key) => tr(_code, key);
}

File: ios/FurniturePlatform/Sources/Networking/LocaleStore.swift
import Foundation

/// Til holati — profil ekranidan tanlanadi, qurilmada saqlanadi. Birinchi
/// marta ochilganda (hali qo'lda tanlanmagan bo'lsa) qurilma tiliga qarab
/// avtomatik aniqlanadi — qo'llab-quvvatlanmasa "O'zbekcha"ga qaytadi.
/// Flutter'dagi `lib/locale_store.dart` bilan bir xil.
@MainActor
final class LocaleStore: ObservableObject {
    private static let prefKey = "fp.locale"

    @Published private(set) var code: String

    init() {
        code = UserDefaults.standard.string(forKey: Self.prefKey) ?? Self.detectDeviceLocale()
    }

    private static func detectDeviceLocale() -> String {
        for preferred in Locale.preferredLanguages {
            let languageCode = Locale(identifier: preferred).language.languageCode?.identifier
            if let languageCode, appLocales.contains(where: { $0.code == languageCode }) {
                return languageCode
            }
        }
        return defaultLocaleCode
    }

    func setLocale(_ code: String) {
        self.code = code
        UserDefaults.standard.set(code, forKey: Self.prefKey)
    }

    func t(_ key: String) -> String { translate(code, key) }
}
